csp_billing_adapter_microsoft/plugin.py

```diff
@@ -81,55 +77,54 @@ def get_account_info(config: Config):
     The information contains the metadata for compute and network.
     """
     account_info = _get_metadata()
-    account_info['compute'] = json.loads(account_info.get('compute', '{}'))
-    account_info['network'] = json.loads(account_info.get('network', '{}'))
     account_info['cloud_provider'] = get_csp_name(config)
 
     return account_info
 
 
-def get_api_header():
-    """
-    Get the header to be used in requests
-
-    Prefer IMDS which requires a token.
-    """
-    request = urllib.request.Request(
-        METADATA_TOKEN_URL,
-        headers={'Metadata': 'True'},
-        method='GET'
-    )
-
+def _get_metadata():
+    metadata = {}
     try:
-        token = urllib.request.urlopen(request).read().decode()
-    except urllib.error.URLError as error:
-        log.error(f'Failed to retrieve metadata token: {str(error)}')
-        return {}
+        metadata = _get_signature()
+        metadata['compute'] = _get_compute_metadata()
+    except ValueError as error:
+        log.error('Could not load JSON from metadata: {}'.format(error))
+        for key in ['compute', 'signature']:
+            if not metadata.get(key):
+                metadata[key] = {}
 
-    return token
+    return metadata
 
 
-def _get_metadata():
-    metadata_options = ['compute', 'network']
+def _get_compute_metadata():
+    compute = json.loads(_fetch_metadata(COMPUTE_URL))
+
     return {
-        metadata_option: _fetch_metadata(metadata_option)
-        for metadata_option in metadata_options
+        'offer': compute.get('offer'),
```

Comment:
Is None a valid value to return for the `offer` field here? Or should we be handling a failure to retrieve this value as an error/exception, in which case we could just use compute['offer'] rather than a get() that defaults to None.

Reply:
I think for a custom image the offer could be none.

Reply:
@jesusbv is my assumption correct about None being valid? If so I guess this can be marked resolved

Reply:
It should not be `None` according to [1] `offer` is always there. But it may be an empty string if the instance is not listed in Marketplace. Here is an example from an openSUSE image that is not in MP

```
python3
>>> import json
>>> import requests
>>> res = requests.get('http://169.254.169.254/metadata/instance?api-version=2020-09-01', headers={'Metadata': 'True'})
>>> metadata = json.loads(res.text)
>>> metadata['compute']['offer']
''
>>>
```

I am not too concerned about he extra function call overhead, and of course it will always be there. As such for all practical purposes this does what is expected. For a safe guard we could add `''` to the get call to make sure we stay consistent with the return type which should always be a string.

@rtamalin I really have no strong preference of `get()` vs `['offer']`


[1] https://learn.microsoft.com/en-us/azure/virtual-machines/instance-metadata-service?tabs=linux